atlasaction/action_test.go

```diff
@@ -0,0 +1,159 @@
+package atlasaction
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"ariga.io/atlas-go-sdk/atlasexec"
+	_ "github.com/mattn/go-sqlite3"
+	"github.com/sethvargo/go-githubactions"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMigrateApply(t *testing.T) {
+	t.Run("local dir", func(t *testing.T) {
+		tt := newT(t)
+		tt.setInput("url", "sqlite://"+tt.db)
+		tt.setInput("dir", "file://testdata/migrations/")
+
+		err := MigrateApply(context.Background(), tt.cli, tt.act)
+		require.NoError(t, err)
+
+		m, err := tt.outputs()
+		require.EqualValues(t, map[string]string{
+			"error": "",
+		}, m)
+	})
+	t.Run("tx-mode", func(t *testing.T) {
+		tt := newT(t)
+		tt.setInput("url", "sqlite://"+tt.db)
+		tt.setInput("dir", "file://testdata/migrations/")
+		tt.setInput("tx-mode", "fake")
+		err := MigrateApply(context.Background(), tt.cli, tt.act)
+
+		// An error here proves that the tx-mode was passed to atlasexec, which
+		// is what we want to test.
+		exp := `run failed: unknown tx-mode "fake"`
+		require.EqualError(t, err, exp)
+		m, err := tt.outputs()
+		require.EqualValues(t, map[string]string{
+			"error": exp,
+		}, m)
+	})
+	t.Run("baseline", func(t *testing.T) {
+		tt := newT(t)
+		tt.setInput("url", "sqlite://"+tt.db)
+		tt.setInput("dir", "file://testdata/migrations/")
```

Comment:
nitpicking a bit, but also a honest question since I had an opposite comment about that in one of my PRs - the raw values here are duplicated several times, why not extract them to a const?

Reply:
I don't have hard rules about when to use a constant and when to reuse a value. Places where it's more obvious to use a constant are when the value is used again downstream in a _comparison_ in which case the const doesn't only save repetition (tho the constant itself is repeated), it proves correctnes (we compare to the same third value which formally must be equal and not prone to typos). 

Another thing to consider is readability. Extracting to a constant sometimes makes the code _less_ readable because you need to jump in the file to understand what that is. In this case, I think its important because the reader can understand which directory I'm referring to. It is very likely that this test func will have many cases with different migration directories to show different use cases.